File: packages/boss-crawler-worker/src/protocol.ts
import { z } from "zod";

export const LoginStartPayloadSchema = z.object({
  source_platform: z.string().optional(),
  executable_path: z.string().optional(),
  user_data_dir: z.string().optional(),
});

export const SessionStatePayloadSchema = z.object({
  cookies: z.any(),
  local_storage: z.any(),
});

export const SearchTaskPayloadSchema = z.object({
  keywords: z.array(z.string().min(1)).min(1),
  source_platform: z.string().optional(),
  filters: z.any().optional().default({}),
  limits: z.any().optional().default({}),
  mode: z.string().optional(),
});

export const CrawlAutoStartPayloadSchema = z.object({
  session: SessionStatePayloadSchema,
  task: SearchTaskPayloadSchema,
  run_id: z.string().optional(),
  user_data_dir: z.string().optional(),
});

export const RefreshJobEvidencePayloadSchema = z.object({
  session: SessionStatePayloadSchema,
  encrypt_job_id: z.string().min(1),
  source_url: z.string().optional(),
  raw_payload: z.any().optional(),
  user_data_dir: z.string().optional(),
});

export const BossMetaSyncPayloadSchema = z.object({
  session: SessionStatePayloadSchema,
});

export const BossChatSyncPayloadSchema = z.object({
  session: SessionStatePayloadSchema,
  limits: z.any().optional().default({}),
});

export const AiAnalyzePayloadSchema = z.object({
  resume_text: z.string().min(1),
  context_text: z.string().optional(),
  resume_files: z.string().optional(),
  job_detail: z.any(),
  filter_reason: z.any().optional(),
  score_reason: z.any().optional(),
  source_context: z.any().optional(),
  review_context: z.any().optional(),
});

export const AiAnalyzeGroupPayloadSchema = z.object({
  context_text: z.string().min(1),
  jobs: z.array(z.any()).min(1),
});

export const AiGreetingPayloadSchema = z.object({
  resume_text: z.string().optional(),
  context_text: z.string().optional(),
  resume_files: z.string().optional(),
  greeting_prompt_extra: z.string().optional(),
  job_detail: z.any(),
  match_report: z.any().optional(),
  filter_reason: z.any().optional(),
  score_reason: z.any().optional(),
  source_context: z.any().optional(),
  review_context: z.any().optional(),
});

export const AiOptimizeResumeForJobPayloadSchema = z.object({
  resume_text: z.string().trim().min(1),
  job_detail: z.any(),
  context_text: z.string().optional(),
  filter_reason: z.any().optional(),
  score_reason: z.any().optional(),
  source_context: z.any().optional(),
  review_context: z.any().optional(),
});

export const AiCompanyScoreBatchPayloadSchema = z.object({
  companies: z
    .array(
      z.object({
        company_name: z.string().trim().min(1),
        jobs_count: z.number().int().nonnegative(),
        source_text: z.string().optional().default(""),
        jobs: z.array(z.any()).optional().default([]),
      }),
    )
    .min(1),
});

export const AiPostCollectionJudgePayloadSchema = z.object({
  profile: z.any().optional(),
  job: z.any(),
  filter_reason: z.any().optional(),
});

export const AiPostCollectionJudgeBatchJobSchema = z.object({
  encrypt_job_id: z.string().trim().min(1),
  job: z.any(),
  filter_reason: z.any().optional(),
});

export const AiPostCollectionJudgeBatchPayloadSchema = z.object({
  profile: z.any().optional(),
  jobs: z.array(AiPostCollectionJudgeBatchJobSchema).min(1),
  concurrency: z.number().int().positive().optional(),
});

export const CommandInSchema = z.union([
  z.object({ type: z.literal("LOGIN_START"), payload: LoginStartPayloadSchema }),
  z.object({
    type: z.literal("CRAWL_AUTO_START"),
    payload: CrawlAutoStartPayloadSchema,
  }),
  z.object({
    type: z.literal("REFRESH_JOB_EVIDENCE"),
    payload: RefreshJobEvidencePayloadSchema,
  }),
  z.object({
    type: z.literal("BOSS_META_SYNC"),
    payload: BossMetaSyncPayloadSchema,
  }),
  z.object({
    type: z.literal("BOSS_CHAT_SYNC"),
    payload: BossChatSyncPayloadSchema,
  }),
  z.object({ type: z.literal("AI_ANALYZE"), payload: AiAnalyzePayloadSchema }),
  z.object({ type: z.literal("AI_ANALYZE_GROUP"), payload: AiAnalyzeGroupPayloadSchema }),
  z.object({ type: z.literal("AI_GREETING"), payload: AiGreetingPayloadSchema }),
  z.object({ type: z.literal("AI_OPTIMIZE_RESUME_FOR_JOB"), payload: AiOptimizeResumeForJobPayloadSchema }),
  z.object({ type: z.literal("AI_COMPANY_SCORE_BATCH"), payload: AiCompanyScoreBatchPayloadSchema }),
  z.object({ type: z.literal("AI_POST_COLLECTION_JUDGE"), payload: AiPostCollectionJudgePayloadSchema }),
  z.object({ type: z.literal("AI_POST_COLLECTION_JUDGE_BATCH"), payload: AiPostCollectionJudgeBatchPayloadSchema }),
  z.object({ type: z.literal("STOP") }),
  z.object({ type: z.literal("PAUSE") }),
  z.object({ type: z.literal("RESUME") }),
]);

export type CommandIn = z.infer<typeof CommandInSchema>;

export const LogPayloadSchema = z.object({
  level: z.string(),
  message: z.string(),
  ts: z.string().optional(),
});

export const ProgressPayloadSchema = z.object({
  keyword: z.string().optional(),
  current_page: z.number().int().nonnegative().optional(),
  captured_job_list: z.number().int().nonnegative().optional(),
  captured_job_detail: z.number().int().nonnegative().optional(),
  filtered_job: z.number().int().nonnegative().optional(),
});

export const LoginStatusPayloadSchema = z.object({
  status: z.string(),
  message: z.string().optional(),
});

export const CookieCollectedPayloadSchema = z.object({
  source_platform: z.string().optional(),
  cookies: z.any(),
  local_storage: z.any(),
});

export const JobListCapturedPayloadSchema = z.object({
  keyword: z.string().optional(),
  filters: z.any().optional(),
  capture_source: z.enum(["natural", "dom_fallback", "api_fallback"]).optional(),
  raw: z.any(),
});

export const JobDetailCapturedPayloadSchema = z.object({
  encrypt_job_id: z.string().min(1),
  zp_data: z.any(),
});

export const JobNormalizedCapturedPayloadSchema = z.object({
  encrypt_job_id: z.string().min(1),
  source_platform: z.string().min(1),
  source_url: z.string().optional(),
  dedup_key: z.string().min(1),
  position_name: z.string().optional(),
  boss_name: z.string().optional(),
  brand_name: z.string().optional(),
  city_name: z.string().optional(),
  salary_desc: z.string().optional(),
  experience_name: z.string().optional(),
  degree_name: z.string().optional(),
  jd_text: z.string().optional(),
  raw_payload: z.any(),
  keyword: z.string().optional(),
  filters: z.any().optional(),
});

export const JobFilteredPayloadSchema = z.object({
  encrypt_job_id: z.string().optional(),
  keyword: z.string().optional(),
  filters: z.any().optional(),
  reason: z.object({
    eligible: z.boolean(),
    blocked_by: z.array(
      z.object({
        rule_type: z.string(),
        field: z.string(),
        value: z.string(),
        reason: z.string(),
      }),
    ),
    matched_preferences: z.array(z.string()),
    missing_preferences: z.array(z.string()),
  }),
  raw: z.any().optional(),
});

export const AiResultPayloadSchema = z.object({
  result: z.any(),
});

export const BossMetaSyncedPayloadSchema = z.object({
  synced_at: z.string().optional(),
  city_group: z.any().optional(),
  filter_conditions: z.any().optional(),
  industry_filter_exemption: z.any().optional(),
});

export const BossChatStatusSyncedPayloadSchema = z.object({
  encrypt_job_id: z.string().min(1),
  communication_status: z.enum(["greeted_unread", "read_no_reply", "replied", "rejected"]),
  boss_name: z.string().optional(),
  brand_name: z.string().optional(),
  position_name: z.string().optional(),
  message_status: z.string().optional(),
  message_preview: z.string().optional(),
  raw_payload: z.any().optional(),
});

export const ErrorPayloadSchema = z.object({
  message: z.string(),
  stack: z.string().optional(),
});

export const EventOutSchema = z.union([
  z.object({ type: z.literal("LOG"), payload: LogPayloadSchema }),
  z.object({ type: z.literal("PROGRESS"), payload: ProgressPayloadSchema }),
  z.object({ type: z.literal("LOGIN_STATUS"), payload: LoginStatusPayloadSchema }),
  z.object({ type: z.literal("COOKIE_COLLECTED"), payload: CookieCollectedPayloadSchema }),
  z.object({ type: z.literal("JOB_LIST_CAPTURED"), payload: JobListCapturedPayloadSchema }),
  z.object({ type: z.literal("JOB_DETAIL_CAPTURED"), payload: JobDetailCapturedPayloadSchema }),
  z.object({ type: z.literal("JOB_NORMALIZED_CAPTURED"), payload: JobNormalizedCapturedPayloadSchema }),
  z.object({ type: z.literal("JOB_FILTERED"), payload: JobFilteredPayloadSchema }),
  z.object({ type: z.literal("AI_RESULT"), payload: AiResultPayloadSchema }),
  z.object({ type: z.literal("BOSS_META_SYNCED"), payload: BossMetaSyncedPayloadSchema }),
  z.object({ type: z.literal("BOSS_CHAT_STATUS_SYNCED"), payload: BossChatStatusSyncedPayloadSchema }),
  z.object({ type: z.literal("FINISHED") }),
  z.object({ type: z.literal("ERROR"), payload: ErrorPayloadSchema }),
]);

export type EventOut = z.infer<typeof EventOutSchema>;

File: src-tauri/src/ipc/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoginStartPayload {
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_platform: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub executable_path: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_data_dir: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionStatePayload {
    pub cookies: Value,
    pub local_storage: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CrawlAutoStartPayload {
    pub session: SessionStatePayload,
    pub task: SearchTaskPayload,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub run_id: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_data_dir: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RefreshJobEvidencePayload {
    pub session: SessionStatePayload,
    pub encrypt_job_id: String,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_url: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub raw_payload: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_data_dir: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BossMetaSyncPayload {
    pub session: SessionStatePayload,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BossChatSyncPayload {
    pub session: SessionStatePayload,
    #[serde(default)]
    pub limits: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchTaskPayload {
    pub keywords: Vec<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_platform: Option<String>,
    #[serde(default)]
    pub filters: Value,
    #[serde(default)]
    pub limits: Value,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mode: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiAnalyzePayload {
    pub resume_text: String,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_text: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resume_files: Option<String>,
    pub job_detail: Value,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filter_reason: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub score_reason: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_context: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub review_context: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiAnalyzeGroupPayload {
    pub context_text: String,
    pub jobs: Vec<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiGreetingPayload {
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resume_text: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_text: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resume_files: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub greeting_prompt_extra: Option<String>,
    pub job_detail: Value,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub match_report: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filter_reason: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub score_reason: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_context: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub review_context: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiCompanyScoreCompanyPayload {
    pub company_name: String,
    pub jobs_count: u32,
    #[serde(default)]
    pub source_text: String,
    #[serde(default)]
    pub jobs: Vec<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiCompanyScoreBatchPayload {
    pub companies: Vec<AiCompanyScoreCompanyPayload>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiPostCollectionJudgePayload {
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub profile: Option<Value>,
    pub job: Value,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filter_reason: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResumeDiagnosePayload {
    pub resume_text: String,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_text: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resume_files: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResumeRewriteModulePayload {
    pub module: String,
    pub resume_text: String,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_text: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resume_files: Option<String>,
    pub module_input: String,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub confirmed_summary: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub confirmed_projects: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub confirmed_experience: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub confirmed_skills: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum CommandIn {
    #[serde(rename = "LOGIN_START")]
    LoginStart(LoginStartPayload),
    #[serde(rename = "CRAWL_AUTO_START")]
    CrawlAutoStart(CrawlAutoStartPayload),
    #[serde(rename = "REFRESH_JOB_EVIDENCE")]
    RefreshJobEvidence(RefreshJobEvidencePayload),
    #[serde(rename = "BOSS_META_SYNC")]
    BossMetaSync(BossMetaSyncPayload),
    #[serde(rename = "BOSS_CHAT_SYNC")]
    BossChatSync(BossChatSyncPayload),
    #[serde(rename = "AI_ANALYZE")]
    AiAnalyze(AiAnalyzePayload),
    #[serde(rename = "AI_ANALYZE_GROUP")]
    AiAnalyzeGroup(AiAnalyzeGroupPayload),
    #[serde(rename = "AI_GREETING")]
    AiGreeting(AiGreetingPayload),
    #[serde(rename = "AI_COMPANY_SCORE_BATCH")]
    AiCompanyScoreBatch(AiCompanyScoreBatchPayload),
    #[serde(rename = "AI_POST_COLLECTION_JUDGE")]
    AiPostCollectionJudge(AiPostCollectionJudgePayload),
    #[serde(rename = "RESUME_DIAGNOSE")]
    ResumeDiagnose(ResumeDiagnosePayload),
    #[serde(rename = "RESUME_REWRITE_MODULE")]
    ResumeRewriteModule(ResumeRewriteModulePayload),
    #[serde(rename = "STOP")]
    Stop,
    #[serde(rename = "PAUSE")]
    Pause,
    #[serde(rename = "RESUME")]
    Resume,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LogPayload {
    pub level: String,
    pub message: String,
    #[serde(default)]
    pub ts: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProgressPayload {
    #[serde(default)]
    pub keyword: Option<String>,
    #[serde(default)]
    pub current_page: Option<u32>,
    #[serde(default)]
    pub captured_job_list: Option<u32>,
    #[serde(default)]
    pub captured_job_detail: Option<u32>,
    #[serde(default)]
    pub filtered_job: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoginStatusPayload {
    pub status: String,
    #[serde(default)]
    pub message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CookieCollectedPayload {
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_platform: Option<String>,
    pub cookies: Value,
    pub local_storage: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobListCapturedPayload {
    #[serde(default)]
    pub keyword: Option<String>,
    #[serde(default)]
    pub filters: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub capture_source: Option<String>,
    pub raw: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobDetailCapturedPayload {
    pub encrypt_job_id: String,
    pub zp_data: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobNormalizedCapturedPayload {
    pub encrypt_job_id: String,
    pub source_platform: String,
    #[serde(default)]
    pub source_url: Option<String>,
    pub dedup_key: String,
    #[serde(default)]
    pub position_name: Option<String>,
    #[serde(default)]
    pub boss_name: Option<String>,
    #[serde(default)]
    pub brand_name: Option<String>,
    #[serde(default)]
    pub city_name: Option<String>,
    #[serde(default)]
    pub salary_desc: Option<String>,
    #[serde(default)]
    pub experience_name: Option<String>,
    #[serde(default)]
    pub degree_name: Option<String>,
    #[serde(default)]
    pub jd_text: Option<String>,
    pub raw_payload: Value,
    #[serde(default)]
    pub keyword: Option<String>,
    #[serde(default)]
    pub filters: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobFilteredPayload {
    #[serde(default)]
    pub encrypt_job_id: Option<String>,
    #[serde(default)]
    pub keyword: Option<String>,
    #[serde(default)]
    pub filters: Option<Value>,
    pub reason: Value,
    #[serde(default)]
    pub raw: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiResultPayload {
    pub result: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BossMetaSyncedPayload {
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub synced_at: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub city_group: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filter_conditions: Option<Value>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub industry_filter_exemption: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BossChatStatusSyncedPayload {
    pub encrypt_job_id: String,
    pub communication_status: String,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub boss_name: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub brand_name: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub position_name: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_status: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_preview: Option<String>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub raw_payload: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorPayload {
    pub message: String,
    #[serde(default)]
    pub stack: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum EventOut {
    #[serde(rename = "LOG")]
    Log(LogPayload),
    #[serde(rename = "PROGRESS")]
    Progress(ProgressPayload),
    #[serde(rename = "LOGIN_STATUS")]
    LoginStatus(LoginStatusPayload),
    #[serde(rename = "COOKIE_COLLECTED")]
    CookieCollected(CookieCollectedPayload),
    #[serde(rename = "JOB_LIST_CAPTURED")]
    JobListCaptured(JobListCapturedPayload),
    #[serde(rename = "JOB_DETAIL_CAPTURED")]
    JobDetailCaptured(JobDetailCapturedPayload),
    #[serde(rename = "JOB_NORMALIZED_CAPTURED")]
    JobNormalizedCaptured(JobNormalizedCapturedPayload),
    #[serde(rename = "JOB_FILTERED")]
    JobFiltered(JobFilteredPayload),
    #[serde(rename = "AI_RESULT")]
    AiResult(AiResultPayload),
    #[serde(rename = "BOSS_META_SYNCED")]
    BossMetaSynced(BossMetaSyncedPayload),
    #[serde(rename = "BOSS_CHAT_STATUS_SYNCED")]
    BossChatStatusSynced(BossChatStatusSyncedPayload),
    #[serde(rename = "FINISHED")]
    Finished,
    #[serde(rename = "ERROR")]
    Error(ErrorPayload),
}
